### DISCUSSION TITLE
Code generation with Textrude

### DISCUSSION THREAD

**Discussion by andbrs** [NeilMacMullen/Textrude#139] [Q&A]
I'm working on a project in which I'm trying to make a code generator for Arduino, I was trying Telosys and Acceleo, but these tools don't convince me completely. I would like to know if you consider that Textrude and scriban have enough capabilities to be able to make this code generator. 
The idea that occurs to me is that for example a person places in the model section something like "servo" and by means of the template, a code like this is generated:

```
#include <Servo.h>
Servo servoMotor;
servoMotor.attach(9);
 // We move to the 0º position
  servoMotor.write(0);
  // We wait 1 second
  delay(1000);
  
  // Move to the 90º position
  servoMotor.write(90);
  // We wait 1 second
  delay(1000);
  
  // We move to position 180º.
  servoMotor.write(180);
  // We wait 1 second
  delay(1000);
```

or something very similar.

**Comment by NeilMacMullen** [comment]
The thing to consider is how much variation you want in the output vs how much is constant.  For example, in your  code above, maybe you want the delay time to be variable or perhaps step by 45 degrees rather than 90?  Generally you want to use the Scriban template as the place to put the "constant" bits of the code as well as the logic to insert parameterised parts and then use  the model to supply parameters.  

If you are intending that the user should be able to generate different kinds of things (perhaps you have an "accelerometer" module as well as a "servo" module) then one way to do it is to create a *library* so that you can write a template that looks something like

 ``` 
#include the library
include "modules.sbn"

#generate servo code using parameters from yaml model..
servo model.servo_delay model.servo_step

#generate accelerometer code...
accelerometer model.accelerometer_sensitivity model.accelerometer_sample_rate

# and so on
```
 

Having said, that Scriban is flexible enough that you can "reflect" over the model so it might be possible to model that defines the structure of the output as well as supplying parameters. 

For example, you could have a yaml model something like

```
modules
 - type: servo
    parameters:
      rate : 100
     step : 45
```
then a template like this...
```
for m in model.modules
 if (m.type == "server")
    rate = m.parameters.rate
    step = m.parameters.step
    ... output template code for servo module
```

**Reply by andbrs** [reply]
It is very interesting what you are telling me, unfortunately, my knowledge is very limited in both YAML and Scriban, could you give me a small functional example of how to do what you are mentioning? Not necessarily with the servo code, just with a line to be able to see your example graphically. Sorry for the inconvenience

**Reply by NeilMacMullen** [reply]
Interesting - I had a quick look at TeloSys and Acceleo.  I think the main advantages of Textrude are: 
- flexibility of model format  
- Scriban looks a little more powerful than some of those languages (I could be mistaken though).
- TextrudeInteractive gives real-time feedback as you editing templates.  Personally, this is what I really find to be the 'killer feature'.

Would genuinely be interested to get your view though; feel free to raise further questions here or on gitter.

**Reply by NeilMacMullen** [reply]
Sure - just download and load the attached project (you will need to unzip it first) 
[cpp_ex.zip](https://github.com/NeilMacMullen/Textrude/files/6267955/cpp_ex.zip)


You should see something like this.... Just change the numbers in the model to see the generated code change.

![image](https://user-images.githubusercontent.com/9131337/113788647-5f212280-9735-11eb-9128-392c43f6b57c.png)

**Reply by andbrs** [reply]
Thank you very much for your help, I really appreciate it. I will look at the code, try to understand it and make some modifications for other types of devices.

**Reply by andbrs** [reply]
That's an excellent example, I've been trying to understand it and it's pretty cool. Now, I was wondering if there is a way to generate the code in a certain place?. For example: I create a parameter called library: `"#include <Servo.h>"` in Yaml and I assign it in the template like this `lib1 = m.parameters.library`, and in the code to call it in the output I used `{{lib1}}`. But unfortunately, Arduino has a structure that I must follow (libraries, void setup () and void loop()). My question is: Is there a way to "call" `{{lib1}} `and can be placed for example above the void setup, or if I need it in the void Loop.

Attached is my modified example.

[cpp_ex.zip](https://github.com/NeilMacMullen/Textrude/files/6268759/cpp_ex.zip)


My final idea is to be able to have the user create for example two "servo" (or other sensors or actuators) in the model and appear twice the code generated for Arduino, in the correct order (libraries, void setup() and void loop()). Can this be done?

Something like this:
```
#include <Servo.h>
Servo servoMotor;
Servo servoMotor2;
 void setup() {
  servoMotor.attach();
  servoMotor2.attach();
}
 
void loop() {
  servoMotor.write();
  delay(1000);
  servoMotor.write();
  delay(1000);
  servoMotor.write();
  delay(1000);

  servoMotor2.write();
  delay(1000);
  servoMotor2.write();
  delay(1000);
  servoMotor2.write();
  delay(1000);
}
```

In this case for two servos, but the user is free to create as many elements as he wants. Can this be done?

**Reply by NeilMacMullen** [reply]
Yes this can be done you just need to accumulate the different sections in variables and output them at the end.  See the attached project...

[cpp_ex.zip](https://github.com/NeilMacMullen/Textrude/files/6271885/cpp_ex.zip)

**Reply by andbrs** [reply]
Thank you very much for your answer, I will analyze it and try some modifications. Thanks again =)

**Reply by andbrs** [reply]
Neil, I was working on the code generator and with the structure, you provided I was able to create several sensors and actuators. Now, I have a question. To save me several lines of code in the void loop I thought about making a class for each of the devices, for example, the servo. I already made a class for the servo in Arduino, which works perfectly.  But I could not make a template in Textrude, which generates these classes and their instances. Could I do something in Textrude like the code below?

The main idea is to create a class for each device and be able to instance from the Yaml model, what the user wants. Is this possible?

```
#include <Servo.h>

class servomotor :
  public  Servo {

  public:
    MyClass() {}

    void pin(uint8_t servoPin) {
      attach(servoPin);
    }

    void servoBackward() {
      for (int val=140;val<=160;val+= 1)
      {write(val); delay (15);}
      for (int val = 160; val >= 140; val -= 1) {
      write(val);
      delay(15);
      }
    }
    void servoBack() {
      for (int val=140;val<=160;val+= 1)
      {write(val); delay (100);}
      for (int val = 160; val >= 140; val -= 1) {
      write(val);
      delay(100);
      }
    }   
};
 
servomotor servo1;
servomotor servo2;

void setup() {
  servo1.pin(9);
  servo2.pin(8);
}

void loop() {

    servo1.servoBackward();
     delay(15);  
    servo2.servoBack();
      delay(15); 
}
```

I worked on this file
[cpp_ex.zip](https://github.com/NeilMacMullen/Textrude/files/6289181/cpp_ex.zip)

**Reply by NeilMacMullen** [reply]
It's exactly the same principle as my previous example used to insert the `#include ` headers..  You just need to use an array to keep track of which classes are needed and then insert them via a variable.   The variable is updated via a function. Here's some skeleton code...

```
{{
#track which classes have been inserted
classList =[]
#variable containing text for classes
classText=""

#use a function to insert the class
func addServoClass
#only do it once
if (array.contains classList "servo") ;ret ;end;
classList = array.add classList "servo"
#body of  servo class
classText = classText +"
class myServo{
... class body
}
"
end

func addServo(s)
#ensure the class  body is added
addServoClass
..do more things
end

#now interate over the users yaml model and insert the code we want
for m in model.modules
if (m.type == "servo")
addServo m
end
end

# insert the acculated classes into the output
classText

}}
```

**Reply by andbrs** [reply]
Thank you very much for your help, I made some modifications, but I can't manage to do as with the previous examples that I can modify the skeleton of the class (I want to add a `def.nameClass`). This is my code so far.

[cpp_ex.zip](https://github.com/NeilMacMullen/Textrude/files/6299373/cpp_ex.zip)



Another question, for the instances of each class how do you recommend me to do it, would it be a new function or in the existing ones?


Thank you very much for your possible help....

**Reply by NeilMacMullen** [reply]
Sorry for the delay - have been a bit busy with "real" life..  I think you are quite close.  Firstly you need to modify the  *addServoClass* function to accept an argument...

```
func addServoClass(def)
```
Then you need to ensure that *def.nameClass* is interpreted as *code*.  As it stands it is just treated as text because it's inside a multiline string.  So change the line to 

```
class " + def.nameClass +":
```
so that *def.nameClass* is treated as a variable and added to the strings to the left and right of itself.

Finally, you need to ensure that the the part of the model that includes the *nameClass* declaration is passed in to the function so change the call to _addServoClass_ in func _AddServo_ to 
```
addServoClass s
```

I've attached a working version.
[cpp_ex.zip](https://github.com/NeilMacMullen/Textrude/files/6319203/cpp_ex.zip)


>Another question, for the instances of each class how do you recommend me to do it, would it be a new function or in the existing ones?

There are lots of ways to arrange this. I used a function per class because it makes for a more obvious grouping but another alternatively would be to use the **include** directive to include text from library scripts.  See the 02_include... example.  You could also use separate models to configure the classes - see the 04_multiple... example for this.

**Reply by andbrs** [reply]
Neil, thank you very much for your response. I will check the example you attached. On another note, I hope you have been able to resolve your busyness with the difficult "real" life. Regards
